src/foreign_key.c

```diff
@@ -0,0 +1,518 @@
+
+	char *conname_addition =
+		ChooseForeignKeyConstraintNameAddition(numfks, confkey, parentRel->rd_id);
+	char *conname = ChooseConstraintName(get_rel_name(fk->conrelid),
+										 conname_addition,
+										 "fkey",
+										 fk->connamespace,
+										 NIL);
+	Oid conoid = CreateConstraintEntry(conname,
+									   fk->connamespace,
+									   CONSTRAINT_FOREIGN,
+									   fk->condeferrable,
+									   fk->condeferred,
+									   fk->convalidated,
+									   fk->oid,
+									   fk->conrelid,
+									   conkey,
+									   numfks,
+									   numfks,
+									   InvalidOid,
+									   indexoid,
+									   chunk->table_id,
+									   mapped_confkey,
+									   conpfeqop,
+									   conppeqop,
+									   conffeqop,
+									   numfks,
+									   fk->confupdtype,
+									   fk->confdeltype,
+#if PG15_GE
+									   confdelsetcols,
+									   numfkdelsetcols,
+#endif
+									   fk->confmatchtype,
+									   NULL,
+									   NULL,
+									   NULL,
+									   false,
+									   1,
+									   false,
+									   false);
+
+	ObjectAddress address, referenced;
+	ObjectAddressSet(address, ConstraintRelationId, conoid);
+	ObjectAddressSet(referenced, ConstraintRelationId, fk->oid);
+	recordDependencyOn(&address, &referenced, DEPENDENCY_INTERNAL);
+
+	CommandCounterIncrement();
+
+	createForeignKeyActionTriggers(fk,
+								   fk->conrelid,
+								   chunk->table_id,
+								   conoid,
+								   indexoid,
+								   parentDelTrigger,
+								   parentUpdTrigger);
+}
+
+/*
+ * Generate the column-name portion of the constraint name for a new foreign
+ * key given the list of column names that reference the referenced
+ * table.  This will be passed to ChooseConstraintName along with the parent
+ * table name and the "fkey" suffix.
+ *
+ * We know that less than NAMEDATALEN characters will actually be used, so we
+ * can truncate the result once we've generated that many.
+ *
+ * This function is based on a function in tablecmds.c in PostgreSQL.
+ */
+static char *
+ChooseForeignKeyConstraintNameAddition(int numkeys, AttrNumber *keys, Oid relid)
+{
+	char buf[NAMEDATALEN * 2];
+	int buflen = 0;
+
+	buf[0] = '\0';
+	for (int i = 0; i < numkeys; i++)
+	{
+		char *name = get_attname(relid, keys[i], false);
+		if (buflen > 0)
+			buf[buflen++] = '_'; /* insert _ between names */
+
+		/*
+		 * At this point we have buflen <= NAMEDATALEN.  name should be less
+		 * than NAMEDATALEN already, but use strlcpy for paranoia.
+		 */
+		strlcpy(buf + buflen, name, NAMEDATALEN);
+		buflen += strlen(buf + buflen);
+		if (buflen >= NAMEDATALEN)
+			break;
+	}
+	return pstrdup(buf);
+}
+
+/*
+ * createForeignKeyActionTriggers
+ *		Create the referenced-side "action" triggers that implement a foreign
+ *		key.
+ */
+static void
+createForeignKeyActionTriggers(Form_pg_constraint fk, Oid relid, Oid refRelOid, Oid constraintOid,
+							   Oid indexOid, Oid parentDelTrigger, Oid parentUpdTrigger)
```

Comment:
Looks like this part is imported from Postgres source code so what about add it to `import` folder?

Reply:
None of these are straight imports, they all had local adjustments and since its only used in this file its more convenient to just have it as static function here.

Reply:
If there are some functions that are mostly copied, let's mention the source function in the comment. I find that sometimes reading the PG equivalent helps to understand what's going on.

Reply:
done